pkg/reconciler/revision/resources/deploy.go

```diff
@@ -137,53 +137,32 @@ func rewriteUserProbe(p *corev1.Probe, userPort int) {
 
 func makePodSpec(rev *v1alpha1.Revision, loggingConfig *logging.Config, tracingConfig *tracingconfig.Config, observabilityConfig *metrics.ObservabilityConfig, autoscalerConfig *autoscaler.Config, deploymentConfig *deployment.Config) (*corev1.PodSpec, error) {
 	queueContainer, err := makeQueueContainer(rev, loggingConfig, tracingConfig, observabilityConfig, autoscalerConfig, deploymentConfig)
-
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue-proxy container: %w", err)
 	}
 
-	userContainer := rev.Spec.GetContainer().DeepCopy()
-	// Adding or removing an overwritten corev1.Container field here? Don't forget to
-	// update the fieldmasks / validations in pkg/apis/serving
-
-	userContainer.VolumeMounts = append(userContainer.VolumeMounts, varLogVolumeMount)
-	userContainer.Lifecycle = userLifecycle
-	userPort := getUserPort(rev)
-	userPortInt := int(userPort)
-	userPortStr := strconv.Itoa(userPortInt)
-	// Replacement is safe as only up to a single port is allowed on the Revision
-	userContainer.Ports = buildContainerPorts(userPort)
-	userContainer.Env = append(userContainer.Env, buildUserPortEnv(userPortStr))
-	userContainer.Env = append(userContainer.Env, getKnativeEnvVar(rev)...)
-	// Explicitly disable stdin and tty allocation
-	userContainer.Stdin = false
-	userContainer.TTY = false
-
-	// Prefer imageDigest from revision if available
-	if rev.Status.ImageDigest != "" {
-		userContainer.Image = rev.Status.ImageDigest
-	}
-
-	if userContainer.TerminationMessagePolicy == "" {
-		userContainer.TerminationMessagePolicy = corev1.TerminationMessageFallbackToLogsOnError
+	containers := []corev1.Container{
+		*queueContainer,
 	}
-
-	if userContainer.ReadinessProbe != nil {
-		if userContainer.ReadinessProbe.HTTPGet != nil || userContainer.ReadinessProbe.TCPSocket != nil {
-			// HTTP and TCP ReadinessProbes are executed by the queue-proxy directly against the
-			// user-container instead of via kubelet.
-			userContainer.ReadinessProbe = nil
+	for i := range rev.Spec.PodSpec.Containers {
+		if len(rev.Spec.PodSpec.Containers[i].Ports) != 0 || len(rev.Spec.PodSpec.Containers) == 1 {
+			servingContainer := makeServingContainer(rev.Spec.GetContainer().DeepCopy(), rev)
+			// Prefer imageDigest from revision if available
+			if rev.Status.ImageDigest != "" {
+				servingContainer.Image = rev.Status.ImageDigest
+			}
+			containers = appendContainer(containers, servingContainer)
```

Comment:
Calling `appendContainer()` here results in O(n^2) time complexity. Probably not an issue since the number of containers are expected to be small.

Can container name uniqueness be done during validation and skipped here?
Alternatively can we use a map here instead of iterating over `containers` checking for name collisions?

Reply:
name check validation not added to check name uniqueness
rather validation added because container name will be unique so it will be easy to decide whether containers need to be appended or not 
I mean the logic of `appendContainer` is to append the containers only if there is change in name.

Reply:
Why is that even needed here? Haven't we made sure there are no name-clashes in validation? If not, we should add that there.

Reply:
Modified thank you :+1: